### PR TITLE
feat(config/presets): add memory caching for config fetching

### PR DESCRIPTION
**Pull request by hasanwhitesource**
## Changes

Added memory caching for preset fetching.

## Context
We agreed that to call `resolveConfigPresets` twice once to resolve the `shallowConfig` and once regularly but doing so will cause double the work for fetches.
- Relates to #15827 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
In which scenarios would the cached result be null?

**Comment by hasanwhitesource** [comment]
Only when the key(preset) isn't cached

**Comment by hasanwhitesource** [comment]
Just a double check but I think it is not needed and it should be just checked if it is undefined.
its undefined only when the cache key (preset) isn't in the cache.

File: lib/config/presets/index.ts [hunk 1]
```diff
@@ -218,15 +219,24 @@ export async function getPreset(
...
+        presetName,
+        tag,
+      })
+    : cachedResult;
   if (!presetConfig) {
     throw new Error(PRESET_DEP_NOT_FOUND);
   }
+  if (is.undefined(cachedResult)) {
+    memCache.set(cacheKey, presetConfig);
+  }
```
Comment (viceice) [inline lib/config/presets/index.ts:238]:
```suggestion
  memCache.set(cacheKey, presetConfig);
```

File: lib/config/presets/index.ts [hunk 2]
```diff
@@ -218,15 +219,24 @@ export async function getPreset(
...
+  const cacheKey = preset;
+  const cachedResult = memCache.get<Preset>(cacheKey);
+  let presetConfig = is.undefined(cachedResult)
+    ? await presetSources[presetSource].getPreset({
+        repo,
+        presetPath,
+        presetName,
+        tag,
+      })
+    : cachedResult;
```
Comment (viceice) [inline lib/config/presets/index.ts:232]:
```suggestion
  const cacheKey = preset;
  const cachedResult = memCache.get<Preset>(cacheKey);
  if (cachedResult) {
    return cachedResult;
  }

  let presetConfig = await presetSources[presetSource].getPreset({
        repo,
        presetPath,
        presetName,
        tag,
      });
```
simpler logic, probably needs prettify

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/presets/index.ts [hunk 3]
```diff
@@ -227,6 +234,7 @@ export async function getPreset(
   if (!presetConfig) {
     throw new Error(PRESET_DEP_NOT_FOUND);
   }
+
```
Comment (rarkins) [inline lib/config/presets/index.ts:237]:
```suggestion
```

File: lib/config/presets/index.ts [hunk 4]
```diff
@@ -218,6 +219,12 @@ export async function getPreset(
   }
   const { presetSource, repo, presetPath, presetName, tag, params } =
     parsePreset(preset);
+
+  const cacheKey = preset;
```
Comment (rarkins) [inline lib/config/presets/index.ts:223]:
```suggestion
  const cacheKey = `preset:${preset}`;
```

Reduce chance of it "colliding" with other uses of the memCache

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by hasanwhitesource** [comment]
Some extra logic needs to be added if we want to cache the result of `getPreset` instead of the fetched preset from the api which is the scope of this pr.
I'm trying to avoid too much changes so we wont have some uncertainties here in this core module.

**Review by rarkins — DISMISSED** [review]

File: lib/config/presets/index.ts [hunk 5]
```diff
@@ -218,15 +219,22 @@ export async function getPreset(
...
+        presetName,
+        tag,
+      })
+    : cachedResult;
   if (!presetConfig) {
     throw new Error(PRESET_DEP_NOT_FOUND);
   }
+  if (is.undefined(cachedResult)) {
+    memCache.set(cacheKey, presetConfig);
+  }
```
Comment (viceice) [inline lib/config/presets/index.ts:237]: Why only cache undefined?

File: lib/config/presets/index.ts [hunk 6]
```diff
@@ -218,15 +219,22 @@ export async function getPreset(
...
+  const cacheKey = `preset:${preset}`;
+  const cachedResult = memCache.get<Preset>(cacheKey);
+  let presetConfig = is.undefined(cachedResult)
+    ? await presetSources[presetSource].getPreset({
+        repo,
+        presetPath,
+        presetName,
+        tag,
+      })
+    : cachedResult;
```
Comment (viceice) [inline lib/config/presets/index.ts:231]: Still don't like that syntax, hard to read
```suggestion
  let presetConfig = memCache.get<Preset>(cacheKey);
  if (is.undefined(cachedResult)) {
    presetConfig = await presetSources[presetSource].getPreset({
        repo,
        presetPath,
        presetName,
        tag,
      });
    }
```

Let's use a little more verbose syntax.

File: lib/config/presets/index.ts (hunk 5 above)
Comment (hasanwhitesource) [reply lib/config/presets/index.ts:237]: Before we were checking if this is null or undefined. I will change to a simple if statement now

Comment (rarkins) [reply lib/config/presets/index.ts:237]: If the previous cache result was undefined (not found) then it caches the new result

Comment (hasanwhitesource) [reply lib/config/presets/index.ts:237]: @viceice Does this look better or keep what we have ?
```  
const { presetSource, repo, presetPath, presetName, tag, params } =
    parsePreset(preset);
  const cacheKey = `preset:${preset}`;
  let presetConfig: Preset | null | undefined = memCache.get<Preset>(cacheKey);
  if (is.nullOrUndefined(presetConfig)) {
    presetConfig = await presetSources[presetSource].getPreset({
      repo,
      presetPath,
      presetName,
      tag,
    });
    if (presetConfig) {
      memCache.set(cacheKey, presetConfig);
    }
  }
  if (!presetConfig) {
    throw new Error(PRESET_DEP_NOT_FOUND);
  }
  logger.trace({ presetConfig }, `Found preset ${preset}`);```

Comment (viceice) [reply lib/config/presets/index.ts:237]: updated my suggestion

File: lib/config/presets/index.spec.ts [hunk 7]
```diff
@@ -307,6 +311,41 @@ describe('config/presets/index', () => {
...
+
+    it('gets preset value from cache when it has been seen', async () => {
+      config.extends = ['github>username/preset-repo'];
+      config.packageRules = [
+        {
+          matchManagers: ['github-actions'],
+          groupName: 'github-actions dependencies',
+        },
+      ];
+      gitHub.getPreset.mockResolvedValue({
```
Comment (viceice) [inline lib/config/presets/index.spec.ts:323]:
```suggestion
      gitHub.getPreset.mockResolvedValueOnce({
```

We want to make sure it only resoved once

File: lib/config/presets/index.spec.ts [hunk 8]
```diff
@@ -307,6 +311,41 @@ describe('config/presets/index', () => {
...
+          {
+            matchDatasources: ['docker'],
+            matchPackageNames: ['ubi'],
+            versioning: 'regex',
+          },
+        ],
+      });
+
+      let res = await presets.resolveConfigPresets(config);
+      res = await presets.resolveConfigPresets(config);
```
Comment (viceice) [inline lib/config/presets/index.spec.ts:334]:
```suggestion
      expect(await presets.resolveConfigPresets(config)).toBeDefined();
      const res = await presets.resolveConfigPresets(config);
```

validate both

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**